test/jdk/javax/swing/JComboBox/bug4166593.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 1998, 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.awt.Robot;
+import java.awt.event.ActionListener;
+import javax.swing.JComboBox;
+import javax.swing.JFrame;
+import javax.swing.JLabel;
+import javax.swing.JPanel;
+import javax.swing.SwingUtilities;
+
+/*
+ * @test
+ * @bug 4166593
+ * @summary Tests that JComboBox fires action events every time the user does an action
+ * @key headful
+ * @run main bug4166593
+ */
+
+public class bug4166593 {
+    static JFrame frame;
+    static JComboBox comboBox;
+    static int numberOfActionEvents = 0;
+
+    public static void main(String[] args) throws Exception {
+        Robot robot = new Robot();
+        SwingUtilities.invokeAndWait(() -> createTestUI());
+        robot.waitForIdle();
+        robot.delay(250);
+
+        // change selected index 3 times
+        SwingUtilities.invokeAndWait(() -> {
+            comboBox.setSelectedIndex(1);
+            comboBox.setSelectedIndex(3);
+            comboBox.setSelectedIndex(2);
+        });
+        robot.waitForIdle();
+        robot.delay(250);
+
+        if (numberOfActionEvents != 3) {
+            throw new RuntimeException("Unexpected number of Action Events!\n" +
+                    "Expected: 3\nActual: " + numberOfActionEvents);
+        }
+
+        if (frame != null) {
```

Comment:
frame is initialized on a different thread and accessed here without proper synchronization

Reply:
added volatile to numberOfActionEvents, moved all access of frame into EDT